compiler/transforms/convert_acc_to_csr.py

```diff
@@ -0,0 +1,254 @@
+        # grab a dict that translates field names to CSR addresses:
+        field_to_csr = dict(self.get_acc(op.accelerator).field_items())
+
+        # emit the llvm assembly code to set csr values:
+        for field, val in op.iter_params():
+            addr = field_to_csr[field]
+            rewriter.insert_op_before_matched_op(
+                [
+                    addr_val := arith.Constant(addr),
+                    llvm.InlineAsmOp(
+                        "csrw $0, $1",
+                        "I, r",
+                        [addr_val, val],
+                        has_side_effects=True,
+                    ),
+                ]
+            )
+        # delete the old setup op
+        rewriter.erase_matched_op(safe_erase=False)
+
+
+class LowerAccLaunchToCsr(LowerAccBasePattern):
+    """
+    Convert launch ops to a single `csr_set $launch_addr, 1`
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: acc.LaunchOp, rewriter: PatternRewriter, /):
+        assert isinstance(op.state.type, acc.StateType)
+        acc_op = self.get_acc(op.state.type.accelerator)
+
+        # insert an op that sets the launch CSR to 1
+        rewriter.replace_matched_op(
+            [
+                addr_val := arith.Constant(acc_op.launch_addr),
+                val := arith.Constant(builtin.IntegerAttr.from_int_and_width(1, 5)),
+                llvm.InlineAsmOp(
+                    "csrw $0, $1",
+                    # I = any 12 bit immediate, K = any 5 bit immediate
+                    # The K allows LLVM to emit an `csrrwi` instruction,
+                    # which has room for one 5 bit immediate only.
+                    "I, K",
+                    [addr_val, val],
+                    has_side_effects=True,
+                ),
+            ],
+            [op.state],
+            safe_erase=False,
+        )
+
+
+class LowerAccAwaitToCsr(LowerAccBasePattern):
+    """
+    Lower await ops to a set of assembly that lowers to a buffer.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: acc.AwaitOp, rewriter: PatternRewriter, /):
+        assert isinstance(op.token.type, acc.StateType | acc.TokenType)
+        acc_op = self.get_acc(op.token.type.accelerator)
+
+        # TODO: this is a temporary solution that will be reworked eventually
+
+        # emit a snax_hwpe-style barrier
+        rewriter.replace_matched_op(
+            [
+                barrier_sw_barrier := arith.Constant(acc_op.barrier_sw_barrier),
+                zero := arith.Constant(builtin.IntegerAttr.from_int_and_width(0, 5)),
+                # FIXME: How to clobber a0?
+                llvm.InlineAsmOp(
+                    (
+                        "\n"
+                        "  csrr a0, $0\n"
+                        "1:\n"
+                        "  bnez a0, 1b\n"
+                        "  csrwi 0x3c5, $1\n"  # Weird clear routine
```

Comment:
We now have a hardcoded CSR address in here, is this what we want?

Reply:
No we should totally get rid of this, the NOPs in this routine are similarly weird and stupid, but it's a dontfix typ of situation for the HWPE accelerator.
If it makes sense we can ask the hardware people to fix it in SNAX